### ISSUE TITLE
Too much magic, not enough explanation

### ISSUE THREAD

**Issue by higginrp** [guildai/guildai#77]
Hi folks,

First off, I really like the project, but **how it works is completely mysterious.**
While I like the approach guild takes to the ML logging problem, you need clearer documentation/ more extensive examples if you are to achieve wisespread adoption.

How it finds hyper-parameters seems discoverable, but how it knows to capture things like **loss**, and **val_acc** in the [image classifier](https://guild.ai/docs/start/image-classifier/) example is entirely unclear. 
Is it making use of some tf.keras magic? Who knows?
Is it making assumptions about how functions/classes are structured and what they return? 
If so, what are these assumptions?

My use case is of similar complexity to the image classifier, albeit implemented in plain Tensorflow rather than tf.keras. The model class definition contains a .train() method, which is called in a train_model function within the script I use to pre-process data and then train the model with.
These methods return, at the end of the training run, the best validation accuracy which guild fails to capture, and despite consulting the documentation at length I am at a loss (no pun intended) on how to fix this.

If you please could answer/provide:
1. What form the scripts should be in, in order for guild to capture values of interest
2. More transparent explanation of how guild infers what to capture.
3. More complex/varied examples of training set-ups and how guild can be configured to work across different set-ups.

Many thanks

**Comment by samedii** [comment]
I agree on the black magic and need for more examples. @gar1t  has been working on new docs for a while now and I expect this is one of the things that will be improved on.

Custom scalars example: https://github.com/guildai/examples/tree/master/custom-scalars

**Comment by gar1t** [comment]
All points taken to heart! As @samedii mentioned, the docs are undergoing a revamp to address these gaps. In response to this issue, I'll port some of these more important topics over to the active docs.

The [custom scalar example](https://github.com/guildai/examples/tree/master/custom-scalars) is a decent reference on how to configure Guild for logging scalars from your script.

Guild's interface to scalar output (or metrics) is via TensorFlow event logs. Guild is independent of TensorFlow and so you don't need to use TensorFlow in any way.

You have two options for logging scalars:

1. Write values to standard output using some well-defined pattern. Guild defaults to the pattern `^NAME: VAL`. You can define whatever you want for your operation. See the example link above for examples.

2. Log the values directly using either a TensorFlow summary writer or [tensorboardX](https://github.com/lanpa/tensorboardX).

The rationale for this approach is that Guild doesn't want to introduce Guild-specific APIs that tie your project code to Guild as a build tool. The cost is that the configuration for these topics is moved outside your project code - or that you have to follow a convention for logging (in this case writing to TF event logs). The upside is that you can layer Guild on top of projects without modifying them. This all gets frustrating when there are gaps in the docs and for that I apologize!

I'll update here when useful docs land.

In the meantime - and I realize this is not a replacement for the need for documentation - if you have specific questions about project configuration, please feel free to post a question to [Slack](http://bit.ly/2VpMrfF). There are a number of examples that cover a variety of use cases and we can help point you in the right direction.

**Comment by gar1t** [comment]
We might keep this issue open forever as I feel it could always apply to Guild! Still waiting on 0.7 to land, but a lot of progress has been made on this front. It's taken WAY too long to get this out but the release is definitely approaching! It's an old fashioned expression, but Real Soon Now.

**Comment by gar1t** [comment]
Finally coming back to this. Thanks @higginrp again for the constructive and fair criticism!

The new docs at [my.guild.ai](https://my.guild.ai/docs) cover these topics in a great deal more detail. There are a number of [cheatsheets](https://my.guild.ai/cheatsheets) that provide sample code and command. Guild now has a number of [examples](https://github.com/guildai/guildai/tree/master/examples) some of which are [annotated](https://my.guild.ai/examples).

The new support site is geared for improved content generation and transparency. If you run into something that you think is too opaque, it would be terrific if you could bring that up. Feel free to update here or open a new issue.

I'll keep this open a while longer in case anyone wants to weigh in. At some point we should close it and address new topics as they come up.
